messages/ruby/lib/cucumber/messages/message/serialization.rb

```diff
@@ -0,0 +1,66 @@
+require 'cucumber/messages/message/utils'
+require 'json'
+
+module Cucumber
+  module Messages
+    class Message
+      include Cucumber::Messages::Message::Utils
+
+      module Serialization
+
+        ##
+        # Returns a new Hash formed from the message attributes
+        # If +camelize:+ keyword parameter is set to true, then keys will be camelized
+        #
+        #   Cucumber::Messages::Duration.new(seconds: 1, nanos: 42).to_h                                 # => { seconds: 1, nanos: 42 }
+        #   Cucumber::Messages::PickleTag.new(name: 'foo', ast_node_id: 'abc-def').to_h(camelize: true)  # => { name: 'foo', astNodeId: 'abc-def' }
+        #
+        # It is recursive so embedded messages are also processed
+        #
+        #   location = Cucumber::Messages::Location.new(line: 2)
+        #   Cucumber::Messages::Comment.new(location: location, text: 'comment').to_h  # => { location: { line: 2, :column: nil }, text: "comment" }
+        #
+
+        def to_h(camelize: false)
```

Comment:
This implementation will return a hash with `nil` values for fields that are `nil`.

This in turn will include all fields in the JSON, which we want to avoid because it increases message size considerably.

Unfortunately `JSON.generate` does not provide any options to not output `nil` values as `null`. We therefore need a `to_hash` implementation that only includes non-nil values.